### PR TITLE
mkcloud: add documentation about ha setups and internal variables

### PR DESCRIPTION
**Pull request by jdsn**

### CODE REVIEW

**Comment by jdsn** [comment]
This PR supersedes https://github.com/SUSE-Cloud/automation/pull/285

**Comment by aspiers** [comment]
LGTM although I would strongly recommend a further commit which changes the usage to use a `<<EOUSAGE` heredoc.

**Comment by jdsn** [comment]
I agree that this is the next logical step. Do you prefer this in the same PR?

**Comment by aspiers** [comment]
I don't care, whichever's faster!

**Comment by jdsn** [comment]
Ok, then I will add the commit to this PR, it makes sense :)

**Comment by jdsn** [comment]
@bmwiedemann @aspiers thanks for (another) review

**Comment by MaximilianMeister** [comment]
@aspiers i don't see the code you are referring to in the timeline. i guess you have to comment in the Files changed tab.

File: scripts/mkcloud [hunk 1]
```diff
@@ -1218,86 +1124,176 @@ function cloudupgrade_reboot_and_redeploy_clients()
...
+        VLAN id for public network
+
+EOUSAGE
     exit 1
+
+# UNDOCUMENTED OPTIONS:
+#
+# virtualcloud
+# cloudfqdn
+# forwardmode
```
Comment (MaximilianMeister) [inline scripts/mkcloud]: @jdsn any chance we could further document the undocumented options in this PR?

**Comment by jdsn** [comment]
@MaximilianMeister click on the commit and there you see the code with the comment

**Comment by MaximilianMeister** [comment]
@jdsn i know, but i dont like that too much. and when you force push, adams comments are lost forever

File: scripts/mkcloud (hunk 1 above)
Comment (jdsn) [inline scripts/mkcloud]: Actually I did :)
But this commit should do the cleanup now. Lets do the rest in another commit and PR, otherwise we will never be able to merge this PR.

Comment (jdsn) [inline scripts/mkcloud]: There is already one PR that changes the old documentation location ....

**Comment by jdsn** [comment]
@MaximilianMeister No, you can click on the outdated diff and still see the comments there.

**Comment by jdsn** [comment]
@MaximilianMeister I did the force push before @aspiers commented, so nothing is lost here.

**Comment by MaximilianMeister** [comment]
@jdsn yes because he did not comment in the commit tab. he commented in the timeline

**Comment by bmwiedemann** [comment]
:+1:

**Comment by MaximilianMeister** [comment]
+1

**Comment by aplanas** [comment]
+1 for my side, but makes sense to read the documentation from an external file, instead of embedding it with the code?

**Comment by MaximilianMeister** [comment]
@aplanas thats a good idea, and we had that already discussed in https://github.com/SUSE-Cloud/automation/pull/228 which read from `docs/mkcloud.md`
i guess this will be one of the next steps for documentation enhancement

**Comment by aspiers** [comment]
> @jdsn i know, but i dont like that too much. and when you force push, adams comments are lost forever

@MaximilianMeister This is not true.  As @jdsn they are sometimes still viewable via the outdated diffs expansion (even after a force push), but sometimes when those vanish (I haven't checked but I suspect they vanish when no hunks from the old commit remain still present in newer commits), you can still access them from the links in e-mail notifications and they are also retrievable from the API.  This is one of the things about GitHub which makes it worse than Gerrit.  [I already blogged about this a year ago](http://blog.adamspiers.org/2013/05/12/in-partial-defense-of-githubs-review-system/), including a link to [a test PR showing some of the good and bad aspects of how it works](https://github.com/aspiers/test/pull/1).

**Comment by aspiers** [comment]
> @aplanas thats a good idea, and we had that already discussed in #228 which read from docs/mkcloud.md
> i guess this will be one of the next steps for documentation enhancement

Maybe although the `usage` now calls `expand_steps` to dynamically generate the expanded steps.  We must be careful not to duplicate the definition of expandable steps in more than one place in the code or docs.

**Comment by MaximilianMeister** [comment]
> you can still access them from the links in e-mail notifications and they are also retrievable from the API. 

@aspiers but maybe only as long as the ref logs are kept (30 days by default i think).
nevertheless i would still prefer to have all relevant discussion in one place

**Comment by jdsn** [comment]
> @MaximilianMeister @aplanas thats a good idea, and we had that already discussed in #228 which read from docs/mkcloud.md
> i guess this will be one of the next steps for documentation enhancement

If our goal is to keep mkcloud as a big beast, then I would agree that it might make sense to have the documentation separately.
But what we currently try to achieve is, to split mklcoud into a set of smaller specialized tools. And these small tools should bring their own documentation with themselves. So the refactoring will also take care, that the documentation gets split off into smaller chunks.

**Comment by aspiers** [comment]
> @aspiers but maybe only as long as the ref logs are kept (30 days by default i think).

Maybe, I have not yet seen any evidence either way.

> Nevertheless i would still prefer to have all relevant discussion in one place

So would I but as long as we continue to use GitHub for reviews, we have to live with this trade-off.  Gerrit doesn't have these problems but it has many other disadvantages.  Personally I would be willing to try http://gerrithub.io/ but I think others are opposed to it, not least because of the legal hoops involved in introducing yet another external tool.
